go/vt/vtgate/engine/route.go

```diff
@@ -751,77 +730,180 @@ func (route *Route) handlePrimary(vcursor VCursor, vindexKeys []sqltypes.Value,
-	}
+
 	for rowNum, vindexKey := range vindexKeys {
-		if len(keyspaceIDs[rowNum]) == 0 {
-			return nil, fmt.Errorf("could not map %v to a keyspace id", vindexKey)
+		if keyspaceIDs[rowNum] == nil {
+			if route.Opcode == InsertSharded {
+				return nil, fmt.Errorf("could not map %v to a keyspace id", vindexKey)
+			}
+			// InsertShardedIgnore: skip the row.
+			continue
 		}
-		bv["_"+colVindex.Column.CompliantName()+strconv.Itoa(rowNum)] = sqltypes.ValueBindVariable(vindexKey)
+		bv[insertVarName(colVindex, rowNum)] = sqltypes.ValueBindVariable(vindexKey)
 	}
 	return keyspaceIDs, nil
 }
 
-func (route *Route) handleNonPrimary(vcursor VCursor, vindexKeys []sqltypes.Value, colVindex *vindexes.ColumnVindex, bv map[string]*querypb.BindVariable, ksids [][]byte) error {
-	if colVindex.Owned {
-		for rowNum, vindexKey := range vindexKeys {
-			if vindexKey.IsNull() {
-				return fmt.Errorf("value must be supplied for column %v", colVindex.Column)
-			}
-			bv["_"+colVindex.Column.CompliantName()+strconv.Itoa(rowNum)] = sqltypes.ValueBindVariable(vindexKey)
+// processOwned creates vindex entries for the values of an owned column for InsertSharded.
+func (route *Route) processOwned(vcursor VCursor, vindexKeys []sqltypes.Value, colVindex *vindexes.ColumnVindex, bv map[string]*querypb.BindVariable, ksids [][]byte) error {
+	for rowNum, vindexKey := range vindexKeys {
+		if vindexKey.IsNull() {
+			return fmt.Errorf("value must be supplied for column %v", colVindex.Column)
+		}
+		bv[insertVarName(colVindex, rowNum)] = sqltypes.ValueBindVariable(vindexKey)
+	}
+	return colVindex.Vindex.(vindexes.Lookup).Create(vcursor, vindexKeys, ksids, false /* ignoreMode */)
+}
+
+// processOwnedIgnore creates vindex entries for the values of an owned column for InsertShardedIgnore.
+func (route *Route) processOwnedIgnore(vcursor VCursor, vindexKeys []sqltypes.Value, colVindex *vindexes.ColumnVindex, bv map[string]*querypb.BindVariable, ksids [][]byte) error {
+	var createIndexes []int
+	var createKeys []sqltypes.Value
+	var createKsids [][]byte
+
+	for rowNum, vindexKey := range vindexKeys {
+		if ksids[rowNum] == nil {
+			continue
 		}
-		err := colVindex.Vindex.(vindexes.Lookup).Create(vcursor, vindexKeys, ksids, false /* ignoreMode */)
+		if vindexKey.IsNull() {
+			return fmt.Errorf("value must be supplied for column %v", colVindex.Column)
+		}
+		createIndexes = append(createIndexes, rowNum)
+		createKeys = append(createKeys, vindexKey)
+		createKsids = append(createKsids, ksids[rowNum])
+		bv[insertVarName(colVindex, rowNum)] = sqltypes.ValueBindVariable(vindexKey)
+	}
+	if createKeys == nil {
+		return nil
+	}
+
+	err := colVindex.Vindex.(vindexes.Lookup).Create(vcursor, createKeys, createKsids, true /* ignoreMode */)
+	if err != nil {
+		return err
+	}
+
+	// After creation, verify that the keys map to the keyspace ids. If not, remove
+	// those that don't map.
+	verified, err := colVindex.Vindex.Verify(vcursor, createKeys, createKsids)
```

Comment:
My concern is adding another round trip. Comparing a pre-vitess insert with the vitess version on a table with one lookup vindex, it goes from one round trip to four (app to vtgate, vtgate insert vindex row, vtgate verify vindex row, vtgate insert real row).

It would be somewhat better if we could parallelize the two inserts (or rather N more generally) but either way avoiding that extra verify step does seem like it'll be a worthy performance improvement in the normal case.